docs/specs/00001-expectation-system.md

```diff
@@ -0,0 +1,150 @@
+# Setting the expectations around codeblock results
+
+## Introduction
+
+When ie executes a codeblock within a markdown document, the successful execution
+of the command alone is usually an indication that we accomplished what the
+document was aiming to achieve, however, it does not make any guarantees about
+the results returned from the execution itself. For example, let's look at a
+codeblock which performs an operation on virtual machines in azure. The
+command may look something like this:
+
+```bash
+az vm create ...
+```
+
+And a successful output to that command may look like:
+
+```json
+{
+  "fqdns": "",
+  "id": "/subscriptions/<subscription-id>/resourceGroups/<resource-group>/providers/Microsoft.Compute/virtualMachines/<vm-name>",
+  "identity": {
+    "systemAssignedIdentity": "<managed-identity>",
+    "userAssignedIdentities": {}
+  },
+  "location": "eastus",
+  "macAddress": "00-0D-3A-1C-6B-66",
+  "powerState": "VM running",
+  "privateIpAddress": "10.0.0.4",
+  "publicIpAddress": "172.178.12.226",
+  "resourceGroup": "<resource-group>",
+  "zones": ""
+}
+```
+
+Looking throughout the output, we can see that there is a significant amount of
+variation that can occur between mutliple successful runs of the previous
+command. However though, most of the time we would only like to specify that the
+output either looks like an output that we've seen before or that a
+certain pattern can be found the output itself. In order for the
+Innovation Engine to be able to solve this problem, we need to create a system
+that allows authors to quantify what they expect the result of a command to
+look like.
+
+## Solution
+
+Most of the codeblocks that the Innovation Engine currently executes are written
+in bash and run executables to carry out tasks on behalf of a user. Instead of
+only relying on the exit code to determine that a codeblock successfully
+executed, we must look at the side effects (output to stdout/stderr) produced
+by the command too. We will only perform comparisons when a command succeeds
+and only from output sent to the standard output stream. If documentation
+authors would like to make comparisons using the comamnd output from standard
+error, they can manually combine the standard output/error stream into the
+output stream within the codeblock itself.
+
+To allow documentation authors to set their expectations around results,
+we would like to introduce a new syntax to markdown documents in the form of
+`expectation` tags. Expectation tags are HTML 5 comments that must be defined
+inside of the markdown documents above a result block and below a codeblock.
+Here's how that would look:
+
+````markdown
+# Example codeblock
+
+```bash
+echo "foo"
+```
+
+<!--expectation type="similarity" value="0.7"-->
```

Comment:
Will we make implementation backwards compatible with the current format? For clarity, I'm not saying this is required, we do not have so many docs that updating them is problematic. However, we will want to fail when encountering an old format if not backward compatible.

Reply:
I initially wanted to, but I think we should just deprecate the old syntax entirely to avoid confusion in the future. It should be easy to update the documents to the new syntax and we could easily add errors when an encountering the older syntax.